src/pip/_internal/download.py

```diff
@@ -181,6 +290,19 @@ def __call__(self, req):
 
         return req
 
+    def _prompt_for_password(self, netloc):
+        # Factored out to allow for easy patching in tests
+        username = six.moves.input("User for %s: " % netloc)
+        password = getpass.getpass("Password: ")
+        return username, password
+
+    def _should_save_password_to_keyring(self):
+        # Factored out to allow for easy patching in tests
+        if not keyring:
+            return False
+        save = six.moves.input("Save credentials to keyring [y/N]: ")
```

Comment:
I would consider using (or at least add a helper function near) `misc.ask()` so we can avoid having too many different ways for invoking `input()`. Also notice the `PIP_NO_INPUT` environment variable checked there.

Reply:
Why does `PIP_NO_INPUT` just raise an exception? Is the idea to catch it and use the default? Why not have `ask` return a default value (or None)?

In particular, you can't safely call `ask_path_exists` right now without checking in advance that `PIP_EXISTS_ACTION` contains a valid value and `PIP_NO_INPUT` is unset, which defeats the point of the helper function.

Reply:
Okay, most of the calls in download.py at least are protected behind `self.prompting`, which is fine. Though the one in `_copy_file` is not.

Reply:
I didn't look closely at `PIP_NO_INPUT` etc. but just think we should be keeping the prompt logic together so it can be handled more consistently in the future (and not get worse / harder to follow).

Reply:
(And yeah, that's why I also suggested adding a helper function near it since `ask()` probably doesn't work for you as is.)